### PR TITLE
Add native commutators to PauliWord and PauliSentence classes

### PR DESCRIPTION
**Pull request by Qottmann**
Adding native commutators for paulis.
Works between `PauliSentence` and any of `Operator, PauliWord, PauliSentence`

Builds on top of https://github.com/PennyLaneAI/pennylane/pull/5051

Todo

- [x] Basic functionality
- [x] Basic functionality tests
- [x] update qml.commutator
- [x] changelog

### CODE REVIEW

**Comment by Qottmann** [comment]
[sc-53927]

**Comment by Qottmann** [comment]
Why native commutator support you ask?

```
n = 10
H1 = PauliWord({0:"X", 1:"Z"})
H2 = PauliWord({0:"X", 1:"Z"})
for i in range(n):
    H1 += 0.5 * PauliWord({i+1:"Y", i+2:"Z"})
    H2 += 0.5 * PauliWord({i+1:"X", i+2:"Y"})
>>> %timeit H1 | H2
253 µs ± 4.32 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
>>> %timeit qml.commutator(H1, H2)
3.12 ms ± 348 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

File: pennylane/pauli/pauli_arithmetic.py [hunk 1]
```diff
@@ -592,6 +631,34 @@ def __truediv__(self, other):
...
+        if len(self) == 0:
+            return copy(other)
+
+        if len(other) == 0:
+            return copy(self)
+
+        for pw1 in self:
+            for pw2 in other:
+                comm_pw, coeff = pw1._commutator(pw2)
+                if comm_pw is not None:
```
Comment (mudit2812) [inline pennylane/pauli/pauli_arithmetic.py:651]: `_commutator` never returns `None`

```suggestion
                if len(comm_pw) != 0:
```

File: pennylane/pauli/pauli_arithmetic.py [hunk 2]
```diff
@@ -592,6 +631,34 @@ def __truediv__(self, other):
...
+                    final_ps[comm_pw] += coeff * self[pw1] * other[pw2]
+
+        return final_ps
+
+    def __or__(self, other):
+        return self.commutator(other)
+
+    def __ror__(self, other):
+        return -1 * qml.commutator(self, other, pauli=True)
```
Comment (mudit2812) [inline pennylane/pauli/pauli_arithmetic.py:660]: Why not this?
```suggestion
        return -1 * self.commutator(other)
```

**Review by mudit2812 — COMMENTED** [review]
Looks good to me! Looks great to me! You can also update `qml.commutator` now to use `PauliSentence.commutator` for `pauli=True`. I'm happy with the current changes and am ready to approve once `qml.commutator` is updated.

File: pennylane/pauli/pauli_arithmetic.py (hunk 1 above)
Comment (Qottmann) [reply pennylane/pauli/pauli_arithmetic.py:651]: Oh good catch! A relic from an old version :)

**Comment by Qottmann** [comment]
Now fully removed the infix commutator, which removes the capability to have the first input be an Operator (e.g. `PauliX(0) | PauliWord({0:"Y"})`, which required some changes to the tests.

**Comment by mudit2812** [comment]
> Why native commutator support you ask?
> 
> ```
> n = 10
> H1 = PauliWord({0:"X", 1:"Z"})
> H2 = PauliWord({0:"X", 1:"Z"})
> for i in range(n):
>     H1 += 0.5 * PauliWord({i+1:"Y", i+2:"Z"})
>     H2 += 0.5 * PauliWord({i+1:"X", i+2:"Y"})
> >>> %timeit H1 | H2
> 253 µs ± 4.32 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
> >>> %timeit qml.commutator(H1, H2)
> 3.12 ms ± 348 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
> ```
> 
> 🚀

Considering the speedup, `qml.comm` should probably be updated to use the native commutators if `Pauli=True`.

**Comment by Qottmann** [comment]
> Considering the speedup, qml.comm should probably be updated to use the native commutators if Pauli=True.

Is there something missing in https://github.com/PennyLaneAI/pennylane/pull/5052/commits/ba5166fd74d3387f5fba4bdbaced7c6776ef68cd ?

**Comment by mudit2812** [comment]
> > Considering the speedup, qml.comm should probably be updated to use the native commutators if Pauli=True.
> 
> Is there something missing in [ba5166f](https://github.com/PennyLaneAI/pennylane/commit/ba5166fd74d3387f5fba4bdbaced7c6776ef68cd) ?

Missed that change. Sorry about that!

File: doc/releases/changelog-dev.md [hunk 3]
```diff
@@ -39,6 +40,16 @@
...
   ```
 
+  We can also compute commutators with Pauli operators natively with the `|` (or) operator.
+  ```pycon
+  >>> op1 = PauliWord({0:"X", 1:"X"})
+  >>> op2 = PauliWord({0:"Y"}) + PauliWord({1:"Y"})
+  >>> op1 | op2
+  2j * Z(0) @ X(1)
+  + 2j * X(0) @ Z(1)
+  ```
```
Comment (mudit2812) [inline doc/releases/changelog-dev.md:50]: This example needs to be updated.

**Review by mudit2812 — COMMENTED** [review]
Thanks Korbinian. I'm ready to approve once test coverage is fixed 😄

File: doc/releases/changelog-dev.md (hunk 3 above)
Comment (Qottmann) [reply doc/releases/changelog-dev.md]: Nice catch, thanks! ✅

**Comment by Qottmann** [comment]
Realized there was a legacy clause in `PauliSentence.commutator` returning `copy(self)` when `len(other)=0`, which doesnt make sense as the empty PauliSentence is the 0 element.

Removed that bit, updated a test checking that the logic is sound now, and updated the docs.

**Comment by Qottmann** [comment]
Ready for review now after some CI trouble.

This PR adds a commutator method native to PauliWord/Sentence instances that is substantially faster.
It also updates`qml.commutator` to use the native method `pauli1.commutator(pauli2)`.

File: pennylane/pauli/pauli_arithmetic.py [hunk 4]
```diff
@@ -312,6 +320,44 @@ def __truediv__(self, other):
...
+    def commutator(self, other):
+        """comm between two PauliWords"""
+        if isinstance(other, PauliWord):
+            new_word, coeff = self._commutator(other)
+            if coeff == 0:
+                return PauliSentence({})
+            return PauliSentence({new_word: coeff})
+
+        if isinstance(other, qml.operation.Operator):
+            op_self = qml.pauli.pauli_sentence(self)
```
Comment (albi3ro) [inline pennylane/pauli/pauli_arithmetic.py:350]: Why not just `PauliSentence({self, 1})`? Or just reproducing some of that logic here?

File: pennylane/pauli/pauli_arithmetic.py [hunk 5]
```diff
@@ -484,6 +530,24 @@ class PauliSentence(dict):
...
+    >>> op1 = PauliWord({0:"X", 1:"X"})
+    >>> op2 = PauliWord({0:"Y"}) + PauliWord({1:"Y"})
+    >>> op1.commutator(op2)
+    2j * Z(0) @ X(1)
+    + 2j * X(0) @ Z(1)
+
+    Or, alternatively, use :func:`~commutator`.
+    >>> qml.commutator(op1, op2, pauli=True)
+
+    Note that we need to specify `pauli=True` as the function returns PennyLane operators by default.
```
Comment (albi3ro) [inline pennylane/pauli/pauli_arithmetic.py:549]:
```suggestion
    Note that we need to specify ``pauli=True`` as :func:`~.commutator` returns PennyLane operators by default.
```

File: pennylane/pauli/pauli_arithmetic.py [hunk 6]
```diff
@@ -312,6 +320,44 @@ def __truediv__(self, other):
...
+        # This may be helpful to developers that need a more lightweight comm between pauli words
+        # without creating PauliSentence classes
+
+        if self.commutes_with(other):
+            return PauliWord({}), 0.0
+        new_word, coeff = self._matmul(other)
+        return new_word, 2 * coeff
+
+    def commutator(self, other):
+        """comm between two PauliWords"""
```
Comment (albi3ro) [inline pennylane/pauli/pauli_arithmetic.py:342]: I think the docstrings for `commutator` are worth fleshing out into our standard full docstring format.

File: pennylane/pauli/pauli_arithmetic.py (hunk 4 above)
Comment (Qottmann) [reply pennylane/pauli/pauli_arithmetic.py]: Yes good point, updated that ✅

File: pennylane/pauli/pauli_arithmetic.py (hunk 6 above)
Comment (Qottmann) [reply pennylane/pauli/pauli_arithmetic.py]: Added ✅ also for PauliSentence.commutator()

**Review by albi3ro — APPROVED** [review]
🎉

**Review by vincentmr — APPROVED** [review]
Nice work @Qottmann . Looks good to me.

**Pull request merged**